### PR TITLE
std.experiment.allocator: use Allocator.expand when available for expandArray

### PR DESCRIPTION
**Pull request by ErnyTech**
Some allocators may have optimized calls for expanding the memory 
block instead of the classic reallocation, so expandArray will use 
Allocator.expand instead of Allocator.reallocate.

### CODE REVIEW

**Comment by thewilsonator** [comment]
That should be a function, rather than 3 duplicated sections of code.

File: std/experimental/allocator/package.d [hunk 1]
```diff
@@ -2263,6 +2263,17 @@ if (isInputRange!R)
     assert(arr == [1, 2, 3, 1, 2, 3, 4]);
 }
 
+private bool implExpandArray(Allocator)(auto ref Allocator alloc, ref void[] array,
+        size_t delta)
+{
+    static if (__traits(hasMember, Allocator, "expand"))
+    {
+        return alloc.expand(buf, T.sizeof * delta);
+    } else {
```
Comment (lesderid) [inline std/experimental/allocator/package.d]: Braces and `else` should be on their own lines (https://dlang.org/dstyle.html#phobos_brackets)

**Comment by shove70** [comment]
@thewilsonator, excuse me, is `unittest` necessary for any fix, enhancement or adjustment? Or when we all think we don't need `unittest`, we don't need to write it? thanks.

**Comment by thewilsonator** [comment]
This change is essentially a fast path optimisation, but one that is provided by the user (or library allocator)in the form of an explicit expand capability (rather than simply reallocate with a larger buffer). The individual components should already be tested and assuming they perform to spec, this change should only result in a performance difference.

So, TL;DR its difficult to test, and a test doesn't provide much value in this case.

**Comment by shove70** [comment]
Thanks, I see  :)

**Pull request closed without merging**